Add database

Warog opened the pull request.

Warog committed:
feature: crate db & create table & create schema.sql & data.sql files

Warog committed:
feature: create a UserDbStorage, add mappers and create CRUD methods

Warog committed:
feature: create a FilmDbStorage, add mappers and create CRUD methods, change a DB scheme

Warog committed:
feature: update SqlRequests, FilmController, validators, change DB scheme!!!, change Film.class parameters

Warog committed:
feature: add a Mpa.class

Warog committed:
fix: Film

Warog committed:
fix: filmUpdate, user&film exceptions,
create test,
change schema,
change data.sql

Warog committed:
fix: add property 'spring.sql.init.continue-on-error=true'

Warog committed:
fix: committed a data.sql,
BECAUSE when creating test data, postman scripts break

Warog committed:
fix: change a data.sql, create test_data.sql,
make a @Sql() in a tests

Alexander-Bolotov commented on src/main/java/ru/yandex/practicum/filmorate/storage/film/FilmDbStorage.java:
```diff
@@ -0,0 +1,214 @@
+import static ru.yandex.practicum.filmorate.storage.SqlRequests.*;
+
+@Slf4j
+@Component("filmDbStorage")
```
Можно использовать @Repository

Alexander-Bolotov commented on src/main/java/ru/yandex/practicum/filmorate/storage/film/InMemoryFilmStorage.java:
```diff
@@ -3,21 +3,22 @@
+import java.util.Map;
 
 @Slf4j
 @Component
```
Можно использовать @Repository

Alexander-Bolotov commented on src/main/java/ru/yandex/practicum/filmorate/storage/user/UserDbStorage.java:
```diff
@@ -0,0 +1,151 @@
+        try {
+            User user = jdbcTemplate.queryForObject(SQL_GET_USER_BY_ID, this::mapRowToUser, id);
+            List<Friend> friends = jdbcTemplate.query(SQL_GET_USER_FRIEND_LIST_BY_ID, this::mapRowToFriend, id);
+            user.setFriends(friends);
```
надо бы проверить user на null

Alexander-Bolotov commented on src/main/java/ru/yandex/practicum/filmorate/storage/user/UserDbStorage.java:
```diff
@@ -0,0 +1,151 @@
+    }
+
+    @Override
+    @Transactional
```
Считается лучше ставить @Transactional над методами сервиса

Alexander-Bolotov commented on src/main/java/ru/yandex/practicum/filmorate/storage/user/UserDbStorage.java:
```diff
@@ -0,0 +1,151 @@
+    @Transactional
+    public List<User> getCommonFriends(long id, long otherId) {
+        List<Long> commonFriendsIds = jdbcTemplate.queryForList(SQL_GET_COMMON_FRIENDS_IDs, Long.class, id, otherId);
+        List<User> commonFriends = commonFriendsIds.stream().map(this::getUser).collect(Collectors.toList());
```
можно сразу перенести в return

Alexander-Bolotov commented on src/main/java/ru/yandex/practicum/filmorate/model/Film.java:
```diff
@@ -8,20 +9,27 @@
     private int id;
     @NotBlank(message = "Навзвание не может быть пустым!")
-    @NotNull(message = "Навзвание фильма не должно быть нулевым!")
+    @NotNull(message = "Название фильма не должно быть нулевым!")
```
@NotBlank включает в себя проверку на null

Alexander-Bolotov commented on src/main/java/ru/yandex/practicum/filmorate/model/Rating.java:
```diff
@@ -0,0 +1,23 @@
+    PG(2),
+    PG_13(3),
+    R(4),
+   NC_17(5);
```
Выравнивание уехало)

Alexander-Bolotov commented on src/main/java/ru/yandex/practicum/filmorate/validation/FilmValidator.java:
```diff
@@ -0,0 +1,59 @@
+@Slf4j
+public class FilmValidator {
+    private static final int DESCRIPTION_MAX_LENGTH = 200;
+    private static final int GENRE_COUNT = 6;
```
нигде не используется

Alexander-Bolotov commented on src/main/java/ru/yandex/practicum/filmorate/validation/FilmValidator.java:
```diff
@@ -0,0 +1,59 @@
+
+    // Валидация данных для фильма
+    public static Film validateFilmInfo(Film film) {
+        if (
```
код стайл

Alexander-Bolotov commented on src/main/java/ru/yandex/practicum/filmorate/validation/FilmValidator.java:
```diff
@@ -0,0 +1,59 @@
+
+
+            log.debug("Валидация фильма прошла успешно!");
+            return film;
```
Обычно в if вкладывается условие при котором выбрасывается исключение, а иначе все Ок и return

Alexander-Bolotov commented on src/main/java/ru/yandex/practicum/filmorate/validation/FriendValidator.java:
```diff
@@ -0,0 +1,23 @@
+            List<Long> friendsOfFriendIDs = jdbcTemplate.query(SQL_GET_USER_FRIEND_LIST_BY_ID, (rs, rowNum) -> rs.getLong("FRIEND_ID"), friendId);
+
+            return friendsOfFriendIDs.stream().anyMatch(friendsOfFriend -> friendsOfFriend == id);
+        }catch(DuplicateKeyException e) {
```
нужен пробел

Warog committed:
fix: change annotation 'Component' on 'Repository' in a DAO classes,
add a NPE checker in DAOs,;
delete '@NotNull' for 'name' parameter on Film.class,
delete unusable parameters,
change,
reorder conditions in 'if'

style: '@Transactional' before the methods,

Warog committed:
fix: change annotation 'Component' on 'Repository' in a DAO classes,
add a NPE checker in DAOs,;
delete '@NotNull' for 'name' parameter on Film.class,
delete unusable parameters,
change,
reorder conditions in 'if'

style: '@Transactional' before the methods,

Warog committed:
fix: change packages names, change a  scheme-image folder
